File: training-data/go/rabin_karp.go
package main

import "fmt"

func rabinKarpSearch(text, pattern string) []int {
	const base = 256
	const mod = 1000000007
	n, m := len(text), len(pattern)
	var matches []int
	if m == 0 || m > n {
		return matches
	}

	var highOrder int64 = 1
	for i := 0; i < m-1; i++ {
		highOrder = (highOrder * base) % mod
	}

	var patternHash, windowHash int64
	for i := 0; i < m; i++ {
		patternHash = (patternHash*base + int64(pattern[i])) % mod
		windowHash = (windowHash*base + int64(text[i])) % mod
	}

	for i := 0; ; i++ {
		if windowHash == patternHash && text[i:i+m] == pattern {
			matches = append(matches, i)
		}
		if i+m >= n {
			break
		}
		windowHash = (windowHash - int64(text[i])*highOrder%mod + mod) % mod
		windowHash = (windowHash*base + int64(text[i+m])) % mod
	}
	return matches
}

func main() {
	fmt.Println(rabinKarpSearch("abxabcabcaby", "abcaby"))
	fmt.Println(rabinKarpSearch("aaaaa", "aa"))
}

File: training-data/rust/rabin_karp.rs
// Rabin-Karp compares rolling hashes instead of raw substrings, so a
// mismatching window is usually rejected in O(1) rather than O(m).
const BASE: u64 = 256;
const MOD: u64 = 1_000_000_007;

fn rabin_karp(text: &str, pattern: &str) -> Vec<usize> {
    let text: Vec<u8> = text.bytes().collect();
    let pattern: Vec<u8> = pattern.bytes().collect();
    let (n, m) = (text.len(), pattern.len());
    if m == 0 || m > n {
        return Vec::new();
    }

    let mut high_order = 1u64;
    for _ in 0..m - 1 {
        high_order = (high_order * BASE) % MOD;
    }

    let mut pattern_hash = 0u64;
    let mut window_hash = 0u64;
    for i in 0..m {
        pattern_hash = (pattern_hash * BASE + pattern[i] as u64) % MOD;
        window_hash = (window_hash * BASE + text[i] as u64) % MOD;
    }

    let mut matches = Vec::new();
    for i in 0..=n - m {
        if window_hash == pattern_hash && &text[i..i + m] == &pattern[..] {
            matches.push(i);
        }
        if i < n - m {
            window_hash = (window_hash + MOD - (text[i] as u64 * high_order) % MOD) % MOD;
            window_hash = (window_hash * BASE + text[i + m] as u64) % MOD;
        }
    }
    matches
}

fn main() {
    println!("{:?}", rabin_karp("abxabcabcaby", "abcaby"));
    println!("{:?}", rabin_karp("aaaaa", "aa"));
    println!("{:?}", rabin_karp("abc", "xyz"));
}
